ethcore/src/blockchain/blockchain.rs

```diff
@@ -612,14 +611,22 @@ impl BlockChain {
 					.add_bloom(&header.log_bloom(), header.number() as usize)
 			},
 			BlockLocation::BranchBecomingCanonChain { ref ancestor, ref route } => {
-				let ancestor_number = self.block_number(ancestor).unwrap();
+				let ancestor_number = try!(self.block_number(ancestor).ok_or(FatalError::MissingBlockDetails { hash: ancestor.clone() }));
 				let start_number = ancestor_number + 1;
 
-				let mut blooms: Vec<H2048> = route.iter()
-					.map(|hash| self.block(hash).unwrap())
-					.map(|bytes| BlockView::new(&bytes).header_view().log_bloom())
+				let blocks: Vec<(H256, Option<Bytes>)> = route.iter()
+					.map(|hash| (hash.clone(), self.block(hash)))
 					.collect();
 
+				let mut blooms: Vec<H2048> = Vec::new();
+
+				for block in blocks.into_iter() {
```

Comment:
That might be usefull:
`One of the keys to collect()'s power is that many things you might not think of as 'collections' actually are. For example, a String is a collection of chars. And a collection of Result<T, E> can be thought of as single [Result<Collection<T>, E>]. See the examples below for more.`
http://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect